archive/map-reduce.md

```diff
@@ -87,25 +106,128 @@ def reduce(key: T, intermediates: Set<Intermediate<T, W>>): Output<K, O>
   do user defined reduce process
 ```
 
-## Data structures on master
+### Data structures on master
+
+The master should keep below data structures.
+
+- The state for each map and reduce tasks.
+- The identity of the worker machine for each map and reduce tasks.
+- The $R$ intermediate files for each completed map tasks. (These files will be piped to the in-progress reduce
+  tasks incrementally).
+
+For more example, the master can store them in json like below.
+
+```json
+{
+  "tasks": [
+    {
+      "task": "map-1",
+      "state": "in-progress",
+      "worker": "worker-1"
+    },
+    {
+      "task": "map-2",
+      "state": "completed",
+      "worker": "worker-2"
+    },
+    {
+      "task": "reduce-1",
+      "state": "idle",
+      "worker": "undefined"
+    }
+  ],
+  "files": [
+    {
+      "location": "/user/var/xxx",
+      "from": "map-2",
+      "to": "reduce-1",
+      "size": "3000"
+    }
+  ]
+}
+```
 
-WIP
+### Fault Tolerance
 
-## Fault Tolerance
+The MapReduce calculation has multiple machines. The distributed computation can be failed.
 
-WIP
+- The worker machine failure
+- The master machine failure
 
-## Locality
+Google prevent such failures using below implementations.
 
-WIP
+- The worker machine failure
+    - The master pings every worker periodically, and consider the worker as failure when it does not response.
+    - Restoration policy
+        - Failed in-progressing reduce/map worker: Re-execute the worker.
+        - Failed completed map worker: Re-execute the worker when the result is in-accessible. The reduce task
+          will be notified if the machine does not read the result of the failed task.
+        - Failed completed reduce worker: No re-execution because it will save result to a global file.
+- The master machine failure
+    - The master saves the status represented by the data structure defined above periodically. This is similar
+      to the database transaction restoration.
+    - Restoration policy: The master loads the status and restore its state.
```

Comment:
This approach could be applied on the stateless MapReduce. Could you also write for stateful MapReduce framework? No need to be detailed, just an introduction is enough. 🙂